tests/functional/kmip/highlevel.spec.js

```diff
@@ -120,4 +122,130 @@ describe('KMIP High Level Driver', () => {
+            `stub for encrypt should have been called twice instead of ${stub.callCount}`);
+    });
+
+    it('should retry and fail the second time on thales session revoked', async () => {
+        const options = {
+            kmip: {
+                providerName: 'tests',
+                client: {
+                    bucketNameAttributeName: null,
+                    compoundCreateActivate: false,
+                },
+                codec: {},
+                transport: {
+                    pipelineDepth: 8,
+                    tls: {
+                        port: 5696,
+                    },
+                },
+            },
+        };
+        const kmipClient = new KMIPClient(options, TTLVCodec, LoopbackServerTransport);
+        // healthcheck to start the connection and the kmipHandshake
+        void await util.promisify(kmipClient.healthcheck.bind(kmipClient))(logger);
+
+        const kmsKey = await util.promisify(kmipClient.createBucketKey.bind(kmipClient))('mybucket', logger);
+
+        void await util.promisify(kmipClient.cipherDataKey.bind(kmipClient))(
+            1, kmsKey, Buffer.from('abc'), logger);
+
+        const reason = 'General Failure';
+        const message = '[NCERRUnauthorizedAccess]: Token has been revoked';
+
+        // the next call should return a thales session revoked error only once
+        const stub = sinon.stub(kmipClient.kmip.transport.channel, '_transform')
+            .callsFake((_cipherFunc, _request, cb) => {
+                kmipClient.kmip.transport.channel.stubbedResult = true;
+                cb(null, kmipClient.kmip.transport.channel.errorResponse(reason, message));
+                kmipClient.kmip.transport.channel.stubbedResult = false;
+            });
+        const tlsWriteSpy = sinon.spy(kmipClient.kmip.transport.channel, 'write');
+        const tlsEndSpy = sinon.spy(kmipClient.kmip.transport.channel, 'end');
+        const tlsConnectSpy = sinon.spy(kmipClient.kmip.transport.channel, 'connect');
+
+
+        await assert.rejects(util.promisify(kmipClient.cipherDataKey.bind(kmipClient))(
+            1, kmsKey, Buffer.from('abc'), logger), (err) => {
+            assert.strictEqual(err.AccessDenied, true);
+            assert.strictEqual(err.description,
+                `KMS (KMIP) error for Encrypt on ${kmsKey}. ${reason}: ${message}`);
+            return true;
+        });
+
+        assert.strictEqual(tlsEndSpy.callCount, 1,
+            `tls connection should have been ended once instead of ${tlsEndSpy.callCount}`);
+
+        assert.strictEqual(tlsConnectSpy.callCount, 1,
+            `tls connect should have been called once instead of ${tlsConnectSpy.callCount}`);
+
+        assert.strictEqual(tlsWriteSpy.callCount, 2,
+            `tls write should have been called twice because of retry instead of ${tlsWriteSpy.callCount}`);
+        assert.strictEqual(Buffer.compare(tlsWriteSpy.getCall(0).args[0], tlsWriteSpy.getCall(1).args[0]), 0,
```

Comment:
Shouldn't retry logic should preserve the original payload for idempotency

Reply:
Yes, this is was test checks

> Compares buf with target and returns a number indicating whether buf comes before, after, or is the same as target in sort order. Comparison is based on the actual sequence of bytes in each Buffer.
>
> 0 is returned if target is the same as buf
> 1 is returned if target should come before buf when sorted.
> -1 is returned if target should come after buf when sorted.
https://nodejs.org/api/buffer.html#bufcomparetarget-targetstart-targetend-sourcestart-sourceend